developer/environment.rst

```diff
@@ -0,0 +1,135 @@
+
+Therefore ``colcon`` provides an alternative way to update the environment.
+
+.dsv files
+~~~~~~~~~~
+
+Since shell scripts can contain arbitrary logic it is opaque from the outside how they affect the environment.
+That makes it impossible to optimize their execution.
+
+Most scripts perform one of the following common operations:
+
+* Set an environment variable to a value.
+* Add a value to an environment variable (using a platform specific delimiter) if the value is not already in the collection.
+* Source / call another script file.
+
+A ``.dsv`` file contains the descriptive information about the intended environment change (instead of shell specific logic).
+The content of such a file uses a semicolon as the delimiter and contains a single line.
+The first value is the ``type`` of the operation followed by a variable number of arguments specific to the operation.
+
+The following list enumerates the supported types and their arguments:
+
+* ``prepend-non-duplicate;<name>;<value>``: Prepend a value ``<value>`` to an environment variable ``<name>`` (using a platform specific delimiter) if the value is not already in the collection.
+  The value is considered to be a path.
+  If the value is not an absolute path the prefix path of the ``.dsv`` file is prepended to the value.
+  An empty value therefore represents the prefix path.
+* ``prepend-non-duplicate-if-exists;<name>;<value>``: Same as ``prepend-non-duplicate`` but only if the path described by the value exists.
+* ``set;<name>;<value>``: Set an environment variable ``<name>`` to a value ``<value>``.
+  If the value is an existing relative path in the install prefix the install prefix is prepended to the value.
+  Otherwise the value is used as is.
+* ``set-if-unset;<name>;<value>``: Same as ``set``  but only if the environment variable is not yet set (or empty).
+* ``source;<path>``: Source / call another script file ``<path>``.
+  If the value is not an absolute path the prefix path of the ``.dsv`` file is prepended.
+
+Implementation
+--------------
+
+Implementing the logic to determine the topological order of packages in every primary shell would be a lot of effort and (depending on the shell) cumbersome.
+Also parsing and interpreting ``.dsv`` files would likely not be much faster then invoking the native scripts.
+
+Therefore both parts are implemented in a Python script located in the root of the install prefix: ``_local_setup_util_<ext>.py``.
+While the Python script itself can't change the environment it is able to efficiently interpret the operations described by the ``.dsv`` files and generate the shell specific commands necessary to update the environment.
+The Python file is templated with information specific to the primary shell it used from hence the ``<ext>`` in the filename.
```

Comment:
Rationale for templating out multiple instances of the script rather than just passing arguments to one?

Reply:
The templated approach was just easier to implement. The script logic only needs to take one set of formatting strings into account.